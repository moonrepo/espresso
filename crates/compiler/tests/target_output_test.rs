mod utils;

use espresso_common::EsTarget;
use espresso_compiler::Compiler;
use espresso_package::Package;
use starbase_sandbox::{assert_snapshot, create_sandbox};
use utils::read_file;

macro_rules! test_target {
    ($method:ident, $target:expr) => {
        #[tokio::test]
        async fn $method() {
            let sandbox = create_sandbox("syntax");
            let package = Package::new(sandbox.path()).unwrap();
            let compiler = Compiler::new(&package).unwrap();
            let out_dir = compiler.compile($target).await.unwrap();

            assert_snapshot!(read_file(out_dir.join("index.mjs")));
        }
    };
}

mod target_output {
    use super::*;

    test_target!(es2015, EsTarget::Es2015);
    test_target!(es2016, EsTarget::Es2016);
    test_target!(es2017, EsTarget::Es2017);
    test_target!(es2018, EsTarget::Es2018);
    test_target!(es2019, EsTarget::Es2019);
    test_target!(es2020, EsTarget::Es2020);
    test_target!(es2021, EsTarget::Es2021);
    test_target!(es2022, EsTarget::Es2022);

    #[tokio::test]
    async fn supports_legacy_decorators() {
        let sandbox = create_sandbox("syntax-legacy-decorators");
        let package = Package::new(sandbox.path()).unwrap();
        let compiler = Compiler::new(&package).unwrap();
        let out_dir = compiler.compile(EsTarget::Es2018).await.unwrap();

        assert_snapshot!(read_file(out_dir.join("index.mjs")));
        assert_snapshot!(read_file(out_dir.join("other.mjs")));
    }
}
