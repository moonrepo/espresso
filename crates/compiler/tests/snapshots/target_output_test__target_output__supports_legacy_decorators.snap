---
source: crates/compiler/tests/target_output_test.rs
expression: "read_file(out_dir.join(\"index.mjs\"))"
---
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _ts_decorate(decorators, target, key, desc) {
    var c1 = arguments.length, r = c1 < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d1;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d1 = decorators[i]) r = (c1 < 3 ? d1(r) : c1 > 3 ? d1(target, key, r) : d1(target, key)) || r;
    return c1 > 3 && r && Object.defineProperty(target, key, r), r;
}
function _ts_metadata(k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
}
function _ts_param(paramIndex, decorator) {
    return function(target, key) {
        decorator(target, key, paramIndex);
    };
}
export let Foo = class Foo {
    method(verbose) {}
    constructor(){
        _define_property(this, "greeting", void 0);
    }
};
_ts_decorate([
    format('Hello, %s'),
    _ts_metadata("design:type", String)
], Foo.prototype, "greeting", void 0);
_ts_decorate([
    c(),
    d('arg'),
    _ts_param(0, required),
    _ts_metadata("design:type", Function),
    _ts_metadata("design:paramtypes", [
        Boolean
    ])
], Foo.prototype, "method", null);
Foo = _ts_decorate([
    a,
    b('arg')
], Foo);

